### DISCUSSION TITLE
Window switcher with window previews

### DISCUSSION THREAD

**Discussion by tokyo4j** [labwc/labwc#2172] [Show and tell]
I made a prototype of window switcher with window previews.
Many things like output scale, output transforms, client viewports are all not considered. Window titles and window icons are future work.

This is the last thing I miss from xfce4, KWin and wayfire so I want to merge this to the master someday, but I'm not sure if the complexity will be acceptable or how we can redesign the configuration.

https://github.com/tokyo4j/labwc/tree/osd-thumbnail

https://github.com/user-attachments/assets/92bd87c0-3230-4948-aa9d-248f787b4a48

**Comment by winteriscariot** [comment]
Love it, similar style of window switcher achievable with lbonn's rofi fork. 

It'd be cool to see mouse interaction with this (and tbh with the default labwc switcher) so that I can click the window I want without having to cycle through to it.

**Comment by johanmalm** [comment]
Amazing. Yes, this is definitely on my wish-list :smile:

Would it be worth considering a modular approach - possibly with function pointers to `render_osd()`. In other words, each OSD module would provide it's own `render_osd()` function to generate the OSD texture.

The benefit of this would be that we don't have to create a hugh number of configuration options to suit different tastes, but that each module would have a certain look/feel/style with only a small number of config options.

Looks like you've written it to show the buffers `live`, so my suggestion might only work if we compositor them onto a still image.

**Comment by tokyo4j** [comment]
Sorry for no reply for a long time.

I added background, window titles and icons:

https://github.com/tokyo4j/labwc/tree/osd-thumbnail

https://github.com/user-attachments/assets/fceb5cc0-b1cf-4ccb-8d26-f888aef24b23

> It'd be cool to see mouse interaction with this (and tbh with the default labwc switcher) so that I can click the window I want without having to cycle through to it.

Yeah it'd be cool. But supporting it for the default window switcher will require a bit of refactoring in `osd.c`, because we're drawing the OSD into a single texture rather than separate scene-nodes.

Another thing I want for window switcher is configurable keys for OSD selection. Currently pressing normal keys or down-arrow key moves the selection forward, and pressing normal key+shift or up-arrow key moves the selection backward. But I want to move the selection via hjkl keys like vim. This is another topic anyway.

> Would it be worth considering a modular approach - possibly with function pointers to render_osd(). In other words, each OSD module would provide it's own render_osd() function to generate the OSD texture.

Yeah that would be a sensible approach. My updated branch (https://github.com/tokyo4j/labwc/tree/osd-thumbnail) splits the implementations of 2 styles of window switchers into `osd-classic.c` and `osd-thumbnails.c`, though the switch for them is currently hardcoded. Each file contains `osd_display_{classic,thumbnails}()` that updates the root scene-node for OSD.

In configuration side, my rough idea of extending theme configuration is like this:
- Rename the existing theme configs for window switcher (`osd.window-switcher.*`) to something like `osd.window-switcher.classic.*`
- Add new configurations in something like `osd.window-switcher.thumbnails.*`

**Comment by Consolatis** [comment]
Nice work!

(This is the first time seeing this)

**Comment by johanmalm** [comment]
Have tested. Works really well.

> Rename the existing theme configs for window switcher (osd.window-switcher.*) to something like osd.window-switcher.classic.*

Yes. Good idea.

**Comment by ffs-akl** [comment]
expect a PR

**Comment by sfs-pra** [comment]
Will OSD be accepted into the main labwc branch or will it exist separately?

**Comment by tokyo4j** [comment]
> expect a PR

> Will OSD be accepted into the main labwc branch or will it exist separately?

I will open a PR later and that will be merged to the master branch. But it'll take some time since I'd like to improve existing window-switcher first (e.g. icons and mouse interactions) so that the new window-switcher proposed here can be fits naturally with the existing codebase or configurations.

**Comment by sfs-pra** [comment]
Packed for Arch, Debian-Ubuntu
If anyone needs them, I can post them
[PKGBUILD.zip](https://github.com/user-attachments/files/18246774/PKGBUILD.zip)

**Comment by fguada** [comment]
I'm not really interested in the previews, but can't wait for icons in osd. This is the last thing I miss from Xfce. Any idea when a PR could be made?

**Reply by tokyo4j** [reply]
That's what I'm planning in this release cycle (0.8.4). Stay tuned!

**Reply by tokyo4j** [reply]
Opened #2621.

**Reply by ffs-akl** [reply]
Are previews next PR?

**Reply by tokyo4j** [reply]
> Are previews next PR?

Yeah I'll work on it after #2620 and #2621 are merged.

**Reply by ffs-akl** [reply]
great work!

**Comment by ghost** [comment]
Will this be included in the 0.8.4 release?

**Reply by tokyo4j** [reply]
No. I haven't opened a PR for it yet, though app icons in the window switcher have been landed in 0.8.4.

Sorry, I'm currently focusing on other stuffs and I cannot guarantee that I'll open a PR in the next release cycle (0.9.0). But I'm not abandoning thumbnails support for window switcher I proposed here.

**Comment by iamloci** [comment]
Absolutely no rush, but are there any updates on when thumbnails will make it into the window switcher?

This is a killer feature for me; it's pretty much the last thing I'm waiting on before fully jumping ship from Xorg.

**Reply by tokyo4j** [reply]
OK, I'll open a draft PR first as soon as possible. Sorry for keeping you await for so long.

**Reply by ffs-akl** [reply]
good news for me

**Reply by iamloci** [reply]
No problem, just fueling my procrastination instead :stuck_out_tongue: 

Much appreciated, awesome project.

**Reply by tokyo4j** [reply]
Opened #2981. Testing and feedback are appreciated.
